### PR TITLE
[Export Refactor][Transformers] Enable loading SparseModels

### PR DESCRIPTION
**Pull request by dbogunowicz**
## Feature Description

The main functionality introduced by this PR enables loading and applying the recipe in the new format to the `text-generation` models. 

The notable changes are:

`src/sparseml/export/export.py` -> allowing the user to provide recipe explicitely to the exporter. If the recipe is not provided, but the `recipe.yaml` is located side-by-side with the model, there will be an attempt to load it automatically

`src/sparseml/pytorch/model_load/helpers.py` -> moving the `log_model_load` out of the `transformers.SparseModel`, otherwise we would get nasty import dependencies. Plus, it just make sense for it to be more general function.

`src/sparseml/transformers/integration_helper_functions.py` -> simplifying the `create_model` function and adapting it to the new changes

`src/sparseml/transformers/utils/helpers.py` -> just adding an extra transformer-specific wrapper around the general `apply_recipe_structure_to_model` method. Just in case we need some transformer-specific functionalities.

`src/sparseml/transformers/utils/initializers.py` -> refactoring `initialize_model`, so now it becomes a comprehensive, main entry point for loading a `SparseModel`.

`src/sparseml/transformers/utils/load_task_model.py` -> look above, this function is an extension of `initialize_model` (now `initialize_sparse_model`)

## Testing

Tests in `tests/sparseml/export/transformers/test_generative_transformers.py` demonstrate the current capabilities of the new `export` module in the context of LLMs.

### CODE REVIEW

File: src/sparseml/transformers/utils/sparse_model.py [hunk 1]
```diff
@@ -452,68 +521,6 @@ def _check_tf(model_name_or_path: str):
...
 
 
-class SparseCausalLM:
-    """
-    Factory class for loading LLMs from the transformers library. Currently OPT and
-    Llama are supported
-    """
-
-    @staticmethod
-    def opt_model_from_pretrained(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/sparse_model.py:462]: check with alex if opt should be separate, otherwise remove.

File: src/sparseml/transformers/utils/sparse_model.py [hunk 2]
```diff
@@ -260,23 +277,32 @@ def text_generation_from_pretrained(
...
+                model=model, model_directory=model_name_or_path, recipe=recipe
+            )
+
         SparseAutoModel.log_model_load(model, model_name_or_path, model_type, delayed)
 
         return model
 
     @staticmethod
-    def text_generation_from_pretrained_distil(
+    def causal_model_from_pretrained_distil(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/sparse_model.py:300]: remove and simply call one function (to be also documented)

File: src/sparseml/transformers/utils/sparse_model.py [hunk 3]
```diff
@@ -260,23 +277,32 @@ def text_generation_from_pretrained(
...
+            torch_dtype=torch_dtype,
         )
+        config = kwargs.get("config")
+        if config:
+            sequence_length = resolve_sequence_length(config)
+            model.seqlen = sequence_length
+        # TODO: Is adding `model.train()` here necessary? The instantiated model
+        # should automatically be in train mode
+        if recipe:
+            apply_structure_to_transformers(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/sparse_model.py:291]: also loads the weights

Comment (dbogunowicz) [reply src/sparseml/transformers/utils/sparse_model.py:291]: assumption is that 'recipy.yaml' exists side-by-side with the model.

Comment (dbogunowicz) [reply src/sparseml/transformers/utils/sparse_model.py:291]: throw a warning if this is not the case.

Comment (dbogunowicz) [reply src/sparseml/transformers/utils/sparse_model.py:291]: if you are training and fine tuning -> train mode,
if you are doing one shot or export -> eval mode,

File: src/sparseml/transformers/utils/load_task_model.py [hunk 4]
```diff
@@ -59,11 +65,13 @@ def load_task_model(
         )
 
     if task in TaskNames.text_generation.value:
-        return SparseAutoModel.text_generation_from_pretrained(
+        return SparseAutoModel.causal_model_from_pretrained(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/load_task_model.py:68]: config loading is here to stay (mandatory argument)

File: src/sparseml/pytorch/model_load/helpers.py [hunk 5]
```diff
@@ -12,164 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/model_load/helpers.py:14]: move log_model_load( out of transformers!

File: src/sparseml/pytorch/model_load/helpers.py [hunk 6]
```diff
@@ -39,6 +39,45 @@
 RECIPE_FILE_NAME = "recipe.yaml"
 
 
+def log_model_load(
```
Comment (dbogunowicz) [inline src/sparseml/pytorch/model_load/helpers.py:42]: This is just moving `log_model_load` from the `SparseModel` class (no reason why it should be folded under that class really, should be a standalone)

File: src/sparseml/transformers/utils/helpers.py [hunk 7]
```diff
@@ -150,3 +182,28 @@ def detect_last_checkpoint(training_args: "TrainingArguments"):  # noqa 821
             )
 
     return last_checkpoint
+
+
+def resolve_sequence_length(config: AutoConfig) -> int:
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/helpers.py:187]: This is just moving this function from another script

File: src/sparseml/transformers/utils/sparse_model.py [hunk 8]
```diff
@@ -250,58 +258,43 @@ def text_generation_from_pretrained(
...
+                **kwargs,
+            )
 
-        return model
+        # TODO: Do we need to call eval() here? Why?
+        model.eval()
+        model.seqlen = sequence_length
 
-    @staticmethod
-    def text_generation_from_pretrained_distil(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/sparse_model.py:275]: Removing this per @markurtz request from yesterday's meeting (@Satrat 's presentation of SparseML Fine-Tuning module)

File: src/sparseml/transformers/utils/sparse_model.py [hunk 9]
```diff
@@ -250,58 +258,43 @@ def text_generation_from_pretrained(
...
-        model = SparseAutoModel.text_generation_from_pretrained(
-            model_name_or_path,
-            model_type="student" if teacher_name_or_path else "model",
-            **model_kwargs,
-        )
-        teacher = (
-            SparseAutoModel.text_generation_from_pretrained(
-                teacher_name_or_path, model_type="teacher", **teacher_kwargs
+        if recipe:
+            apply_structure_to_transformers(
```
Comment (dbogunowicz) [inline src/sparseml/transformers/utils/sparse_model.py:291]: @bfineran this function both applies the structure and loads weight, confirmed by @Satrat

File: src/sparseml/export/export.py [hunk 10]
```diff
@@ -109,14 +113,17 @@ def export(
     :param task: Optional task to use for exporting the model.
         Defaults to None.
     """
+    if task is not None:
+        task = task.replace("_", "-").replace(" ", "-")
+    if integration is not None:
+        integration = integration.replace("_", "-").replace(" ", "-")
```
Comment (Satrat) [inline src/sparseml/export/export.py:119]: nit: helper fn?

File: src/sparseml/pytorch/model_load/helpers.py (hunk 6 above)
Comment (Satrat) [reply src/sparseml/pytorch/model_load/helpers.py:42]: nice!

File: src/sparseml/transformers/integration_helper_functions.py [hunk 11]
```diff
@@ -74,7 +78,10 @@ def create_model(
 
     if task is None:
         raise ValueError("To create a transformer model, a task must be specified")
-    task = task.replace("_", "-")
+    if recipe is None:
+        _LOGGER.info(f"Attempting to apply default recipe: {RECIPE_NAME}")
+    else:
+        _LOGGER.info(f"Attempting to apply user-specified recipe: {recipe}")
```
Comment (Satrat) [inline src/sparseml/transformers/integration_helper_functions.py:84]: lets clean up this logic to something like this:
```
if default_recipe_exists and recipe_provided:
     warn that we're overwriting the model's input recipe, probably a bad idea
      use recipe_provided
elif recipe_provided:
    warn that this recipe should have been previously applied to the model
    use recipe_provided
elif default_recipe:
    just the info log needed
    use default_recipe
else: # neither recipe exists
    warn that no recipe was applied
     carry on
```

File: src/sparseml/transformers/utils/sparse_model.py [hunk 12]
```diff
@@ -250,58 +258,43 @@ def text_generation_from_pretrained(
...
-            **kwargs,
-        )
-        SparseAutoModel.log_model_load(model, model_name_or_path, model_type, delayed)
+            model = OPTForCausalLM.from_pretrained(
+                model_name_or_path,
+                torch_dtype=torch_dtype,
+                config=config,
+                trust_remote_code=trust_remote_code,
+                **kwargs,
+            )
```
Comment (Satrat) [inline src/sparseml/transformers/utils/sparse_model.py:276]: It should be safe to change this to the automodel, I'm just not sure if these torch.nn..... = skip lines can be deleted or not

File: src/sparseml/transformers/utils/sparse_model.py [hunk 13]
```diff
@@ -250,58 +258,43 @@ def text_generation_from_pretrained(
...
 
-        return model
+        # TODO: Do we need to call eval() here? Why?
+        model.eval()
```
Comment (Satrat) [inline src/sparseml/transformers/utils/sparse_model.py:287]: Could we make eval vs train an argument? It depends on if we're doing training(train) or oneshot/export(eval)

File: tests/sparseml/transformers/obcq/test_obcq.py [hunk 14]
```diff
@@ -25,7 +25,7 @@
 from sparseml.transformers.data import TransformersDataset
 from sparseml.transformers.sparsification.obcq.obcq import one_shot
 from sparseml.transformers.sparsification.obcq.utils.helpers import llama_forward
-from sparseml.transformers.utils.model import SparseCausalLM
+from sparseml.transformers.utils.sparse_model import SparseCausalLM
```
Comment (Satrat) [inline tests/sparseml/transformers/obcq/test_obcq.py:28]: I think we can remove this and just use the SparseAutoModel?

File: src/sparseml/transformers/sparsification/obcq/obcq.py [hunk 15]
```diff
@@ -34,7 +34,7 @@
     llama_forward,
     opt_forward,
 )
-from sparseml.transformers.utils.model import SparseCausalLM
+from sparseml.transformers.utils.sparse_model import SparseCausalLM
```
Comment (Satrat) [inline src/sparseml/transformers/sparsification/obcq/obcq.py:37]: Was there a reason not to replace this with the SparseAutoModel? Other than the OPT edge case, I think we only need to add in precision support

**Comment by Satrat** [comment]
Oh one other issue I see: the new `text_generation.py` v2 finetuning script uses the pretrained_from_distil function, that needs to be updated

File: src/sparseml/export/export.py (hunk 10 above)
Comment (bfineran) [reply src/sparseml/export/export.py:119]: +1, this should ideally be resolved at the registry mixin level (all lower case, remove special characters and whitespace)

File: src/sparseml/transformers/utils/helpers.py [hunk 16]
```diff
@@ -150,3 +182,28 @@ def detect_last_checkpoint(training_args: "TrainingArguments"):  # noqa 821
...
+
+    elif hasattr(config, "max_seq_len"):
+        sequence_length = config.max_seq_len
+    else:
+        raise ValueError(
+            "Could not infer a default sequence length "
+            "from the HF transformers config. Please specify "
+            "the sequence length with --sequence_length"
+        )
+    _LOGGER.info(
```
Comment (bfineran) [inline src/sparseml/transformers/utils/helpers.py:205]: should be debug for any non critical logs on our regular path

File: src/sparseml/transformers/utils/initializers.py [hunk 17]
```diff
@@ -73,39 +74,60 @@ def initialize_tokenizer(
...
     config: AutoConfig,
     trust_remote_code: bool = False,
+    recipe: Optional[Union[str, Path]] = None,
     device: Optional[str] = None,
+    **model_kwargs,
 ) -> AutoModel:
     """
     Initialize a model from a given path
 
+    Initialize a sparse model from a given path. This will:
```
Comment (bfineran) [inline src/sparseml/transformers/utils/initializers.py:99]: let's include a usage example here

**Review by bfineran — COMMENTED** [review]
could you update the PR description to include an example of `initialize_sparse_model`

File: src/sparseml/export/export.py (hunk 10 above)
Comment (dbogunowicz) [reply src/sparseml/export/export.py:119]: Please take a look at: https://github.com/neuralmagic/sparsezoo/pull/404
It's been sitting for a while in the "purgatory", but has already been reviewed once by Sara.
@Satrat , @bfineran if you approved this sparsezoo feature, we could remove the need for unifiying the `integration` string (`RegistryMixin` will take care of it). For the task itself, I have also a good solution, will implement in a second.

File: src/sparseml/transformers/utils/sparse_model.py (hunk 12 above)
Comment (dbogunowicz) [reply src/sparseml/transformers/utils/sparse_model.py:276]: @alexm-nm can we have your buy-in here?

File: src/sparseml/transformers/utils/sparse_model.py (hunk 13 above)
Comment (dbogunowicz) [reply src/sparseml/transformers/utils/sparse_model.py:287]: Can we just set this appropriately in the code paths themselves? I feel like adding a boolean argument would be a bit weird. It's just more verbose and informative to call `model.eval()` or `model.train()` explicitly in the code.

**Review by bfineran — APPROVED** [review]
LGTM - is the intention to hand off `initialize_sparse_model` to MLR instead of the previously discussed SparseAutoModel?

**Comment by dbogunowicz** [comment]
Yes

File: src/sparseml/transformers/sparsification/obcq/obcq.py [hunk 18]
```diff
@@ -91,18 +91,18 @@ def one_shot(
...
     elif "llama" in model_type or "mistral" in model_type:
-        model_loader_fn = SparseCausalLM.auto_model_from_pretrained
+        model_loader_fn = SparseAutoModel.text_classification_from_pretrained
         forward_fn = llama_forward
     else:
         _LOGGER.warning(
             f"A supported model type({SUPPORTED_MODELS}) could not be "
             f"parsed from model_path={model_path}. Defaulting to "
-            "AutoModelForCausalLM loading. "
+            "SparseAutoModel loading. "
```
[the remark this answers is not in the document]
Comment (Satrat) [reply src/sparseml/transformers/sparsification/obcq/obcq.py:103]: I think we can get rid of the if statement altogether, and just condition on the model for setting `forward_fn`

File: src/sparseml/transformers/utils/initializers.py [hunk 19]
```diff
@@ -186,6 +201,68 @@ def initialize_trainer(
...
+) -> Union[str, Path, None]:
+    """
+    Resolve the recipe to apply to the model.
+    If the recipe is None, will look for a recipe in the model_path
+
+    :param recipe: the recipe to apply to the model.
+        If None, will look for a recipe in the model_path
+    :param model_path: the path to the model to load
+    :return: the resolved recipe
+    """
```
Comment (Satrat) [inline src/sparseml/transformers/utils/initializers.py:215]: Nice, the logging messages are really clear, and good to have this in a helper function

**Review by Satrat — APPROVED** [review]
LGTM, had one nitpick on the obcq script. Should we update the OBCQ and finetuning scripts to use the new initialize_sparse_model funcion? That can also be a later PR though
